{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import platform as python_platform\n",
    "\n",
    "class Platform:\n",
    "    WINDOWS = \"Windows\"\n",
    "    MACOS = \"Darwin\"\n",
    "\n",
    "platform = python_platform.system()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "windows_root_directory = \"D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling\"\n",
    "macos_root_directory = \"???\"\n",
    "\n",
    "root_directory = windows_root_directory if platform == Platform.WINDOWS else macos_root_directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "from selenium.webdriver.common.by import By\n",
    "\n",
    "import sys\n",
    "sys.path.append(root_directory)\n",
    "\n",
    "import os\n",
    "from os.path import join, getsize\n",
    "import threading\n",
    "import time\n",
    "import subprocess\n",
    "import signal\n",
    "\n",
    "import re\n",
    "\n",
    "import file_utils as fu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json\n"
     ]
    }
   ],
   "source": [
    "json_dir_path = fu.relative_to_absolute_path(\n",
    "    \"GetAllSongHTML/song_list_link_by_artist/json\",\n",
    "    root_path=root_directory\n",
    ")\n",
    "\n",
    "print(json_dir_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'a': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/a.json', 'b': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/b.json', 'c': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/c.json', 'd': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/d.json', 'e': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/e.json', 'f': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/f.json', 'g': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/g.json', 'h': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/h.json', 'i': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/i.json', 'j': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/j.json', 'k': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/k.json', 'l': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/l.json', 'm': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/m.json', 'n': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/n.json', 'o': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/o.json', 'p': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/p.json', 'q': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/q.json', 'r': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/r.json', 's': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/s.json', 't': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/t.json', 'u': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/u.json', 'v': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/v.json', 'w': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/w.json', 'x': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/x.json', 'y': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/y.json', 'z': 'D:/Vector A/0. KHTN/Nam 4/HKII/Thesis/Brainstorming/DataCrawling/GetAllSongHTML/song_list_link_by_artist/json/z.json'}\n"
     ]
    }
   ],
   "source": [
    "song_links_by_alphabet = fu.get_file_path_list_in_dir(json_dir_path)\n",
    "\n",
    "song_links_by_alphabet = {\n",
    "    path.split(\"/\")[-1].split(\".\")[0].lower() : path\n",
    "    for path in song_links_by_alphabet\n",
    "}\n",
    "\n",
    "print(song_links_by_alphabet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a/a-boogie-wit-da-hoodie/drowning.html\n"
     ]
    }
   ],
   "source": [
    "def song_link_to_relative_html_file_path(link):\n",
    "    artist_letter = link.split(\"/\")[-2][0]\n",
    "    artist = link.split(\"/\")[-2]\n",
    "    song = link.split(\"/\")[-1]\n",
    "\n",
    "    return f\"{artist_letter}/{artist}/{song}.html\"\n",
    "\n",
    "\n",
    "print(\n",
    "    song_link_to_relative_html_file_path(\n",
    "        fu.read_data_from_json_file(\n",
    "            song_links_by_alphabet[\"a\"])[0][\"link\"]\n",
    "        )\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "webdriver_pool = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scroll_to_bottom(webdriver_pool_index, scroll_pause_time = 0.5):\n",
    "    # Get the screen height of the web\n",
    "    screen_height = webdriver_pool[webdriver_pool_index][\"browser\"].execute_script(\"return window.screen.height;\") \n",
    "    i = 1\n",
    "\n",
    "    max_scroll_times = 25\n",
    "\n",
    "    while True:\n",
    "        print(f\"Scrolled: {i} time(s)\")\n",
    "\n",
    "        # Scroll one screen height each time\n",
    "        try:\n",
    "            webdriver_pool[webdriver_pool_index][\"browser\"].execute_script(f\"window.scrollTo(0, {screen_height}*{i});\")\n",
    "        except:\n",
    "            print(f\"Error while scrolling\")\n",
    "            raise Exception(\"Error while scrolling\")\n",
    "        \n",
    "        time.sleep(scroll_pause_time)\n",
    "        \n",
    "        # Cập nhật scroll_height sau mỗi lần scroll, vì scroll_height có thể thay đổi sau khi scroll trang\n",
    "        scroll_height = webdriver_pool[webdriver_pool_index][\"browser\"].execute_script(\"return document.body.scrollHeight;\")\n",
    "\n",
    "        i += 1\n",
    "        if (screen_height) * i > scroll_height:\n",
    "            break\n",
    "\n",
    "        if i > max_scroll_times:\n",
    "            print(f\"Max scroll times reached\")\n",
    "            raise Exception(\"Max scroll times reached\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scroll_and_get_html_content(url, webdriver_pool_index):\n",
    "    try:\n",
    "        webdriver_pool[webdriver_pool_index][\"browser\"].get(url)\n",
    "    except:\n",
    "        print(f\"Failed to get {url}\")\n",
    "        raise Exception(f\"Failed to get {url}\")\n",
    "    \n",
    "    scroll_to_bottom(webdriver_pool_index)\n",
    "    \n",
    "    contents = webdriver_pool[webdriver_pool_index][\"browser\"].page_source\n",
    "\n",
    "    return contents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_song_raw_html(url, webdriver_pool_index):\n",
    "    print(f\"{url}: processing...\")\n",
    "    \n",
    "    page_source = scroll_and_get_html_content(url, webdriver_pool_index)\n",
    "\n",
    "    path_to_save_html = fu.relative_to_absolute_path(\n",
    "        song_link_to_relative_html_file_path(url),\n",
    "        root_path=fu.relative_to_absolute_path(\n",
    "            \"GetAllSongHTML/song_list/raw_html\",\n",
    "            root_path=root_directory\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    os.makedirs(os.path.dirname(path_to_save_html), exist_ok=True)\n",
    "\n",
    "    fu.write_data_to_html_file(page_source, path_to_save_html)\n",
    "\n",
    "    print(f\"{url}: done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_edge_in_remote_debugging_mode(port, platform, browser_instance_data_dir):\n",
    "    print(f\"Opening Edge in remote debugging mode on port {port}...\")\n",
    "\n",
    "    macos_command = f'''open -na \"Microsoft Edge.app\" --args\n",
    "                        --remote-debugging-port={port}\n",
    "                        --user-data-dir=\"{browser_instance_data_dir}\"'''\n",
    "\n",
    "    windows_command = f'''start msedge.exe \n",
    "                        --remote-debugging-port={port} \n",
    "                        --user-data-dir=\"{browser_instance_data_dir}'''\n",
    "\n",
    "    command = macos_command if platform == Platform.MACOS else windows_command\n",
    "    \n",
    "    os.system(command.replace(\"\\n\", \" \"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def init_selenium_in_remote_debugging_mode(port):\n",
    "    options = webdriver.EdgeOptions()\n",
    "    options.add_experimental_option(\"debuggerAddress\", f\"localhost:{port}\")\n",
    "\n",
    "    # browser.maximize_window()\n",
    "    \n",
    "    webdriver_pool.append(\n",
    "        {\n",
    "            \"port\": port,\n",
    "            \"browser\": webdriver.Edge(options=options),\n",
    "            \"browser_status\": BrowserStatus.AVAILABLE\n",
    "        }\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def windows_get_PID_of_process_running_on_port(port):\n",
    "    command = f\"netstat -a -n -o | findstr :{port}\"\n",
    "\n",
    "    result = subprocess.run(command, shell=True, capture_output=True, text=True)\n",
    "\n",
    "    #   TCP    127.0.0.1:40000        0.0.0.0:0              LISTENING       6920\n",
    "    regex_pattern = r\"\\s+TCP\\s+127.0.0.1:\\d+\\s+\\d+.\\d+.\\d+.\\d+:\\d+\\s+LISTENING\\s+(.+)\"\n",
    "\n",
    "    match = re.search(regex_pattern, result.stdout)\n",
    "\n",
    "    pid = None\n",
    "\n",
    "    if match:\n",
    "        pid = match.group(1).replace(\" \", \"\")\n",
    "    else:\n",
    "        pid = None\n",
    "\n",
    "    return pid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kill_process_running_on_port(port, platform=Platform.WINDOWS):\n",
    "    print(f\"Killing process running on port {port}...\")\n",
    "\n",
    "    try:\n",
    "        if platform == Platform.MACOS:\n",
    "            result = os.system(f\"lsof -ti tcp:{port} | xargs kill -9\")\n",
    "        else:\n",
    "            pid = windows_get_PID_of_process_running_on_port(port)\n",
    "\n",
    "            if pid != None:\n",
    "                result = subprocess.run(\n",
    "                    f\"taskkill /PID {pid} /F\", \n",
    "                    shell=True, \n",
    "                    capture_output=True, \n",
    "                    text=True\n",
    "                ).stdout\n",
    "            else:\n",
    "                result = f\"No process is running on port {port}\"\n",
    "    except:\n",
    "        print(f\"Error when killing process running on port {port}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "class BrowserStatus:\n",
    "    AVAILABLE = \"available\"\n",
    "    BUSY = \"busy\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def start_webdriver(port, platform):\n",
    "    # Start webdriver if port is not in the pool\n",
    "    if len(webdriver_pool) == 0 or port not in [webdriver[\"port\"] for webdriver in webdriver_pool]:\n",
    "        kill_process_running_on_port(port, platform)\n",
    "\n",
    "        open_edge_in_remote_debugging_mode(\n",
    "            port,\n",
    "            platform,\n",
    "            fu.relative_to_absolute_path(\n",
    "                f\"GetAllSongHTML/song_list/browser_instance_data/{port}\",\n",
    "                root_path=root_directory\n",
    "            )\n",
    "        )\n",
    "\n",
    "        init_selenium_in_remote_debugging_mode(port)\n",
    "    else:\n",
    "        print(f\"Webdriver on port {port} is already started\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def init_webdriver_pool(n_webdriver=5):\n",
    "\n",
    "    webdriver_pool.clear()\n",
    "\n",
    "    if platform == Platform.WINDOWS:\n",
    "        os.system(\"taskkill /F /IM msedgedriver.exe\")\n",
    "        os.system(\"taskkill /F /IM msedge.exe\")\n",
    "\n",
    "    for i in range(n_webdriver):\n",
    "        port = 40000 + i\n",
    "\n",
    "        kill_process_running_on_port(port, platform)\n",
    "        \n",
    "        start_webdriver_thread = threading.Thread(\n",
    "            target=start_webdriver, \n",
    "            args=(port, platform, )\n",
    "        )\n",
    "        \n",
    "        start_webdriver_thread.start()\n",
    "\n",
    "        start_webdriver_thread.join(15)\n",
    "\n",
    "        # Check if the webdriver is started successfully\n",
    "        if len(webdriver_pool) != i + 1:\n",
    "            print(f\"Failed to start webdriver on port {port}. Retrying...\")\n",
    "            i -= 1\n",
    "        else:\n",
    "            print(f\"Webdriver on port {port} started successfully\")           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def restart_webdriver_pool():\n",
    "    n_webdriver_item = len(webdriver_pool)\n",
    "\n",
    "    for webdriver_instance in webdriver_pool:\n",
    "        kill_process_running_on_port(webdriver_instance[\"port\"], platform)\n",
    "\n",
    "    init_webdriver_pool(n_webdriver_item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def change_port_status(port, status):\n",
    "    for i in range(len(webdriver_pool)):\n",
    "        if webdriver_pool[i][\"port\"] == port:\n",
    "            webdriver_pool[i][\"browser_status\"] = status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_raw_html_with_webdriver_from_pool(url, webdriver_pool_index):\n",
    "    if (webdriver_pool[webdriver_pool_index][\"browser_status\"] == BrowserStatus.AVAILABLE):\n",
    "        try:\n",
    "            change_port_status(webdriver_pool[webdriver_pool_index][\"port\"], BrowserStatus.BUSY)\n",
    "            get_song_raw_html(url, webdriver_pool_index)\n",
    "            change_port_status(webdriver_pool[webdriver_pool_index][\"port\"], BrowserStatus.AVAILABLE)\n",
    "        except:\n",
    "            restart_webdriver_pool()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_song_raw_html_by_alphabet(letter, song_per_batch=5):\n",
    "    # Lấy danh sách link của letter\n",
    "    song_links = fu.read_data_from_json_file(song_links_by_alphabet[letter])\n",
    "\n",
    "    # Xóa các song đã được crawl\n",
    "    song_links = [song for song in song_links\n",
    "        if not os.path.exists(\n",
    "            fu.relative_to_absolute_path(\n",
    "                song_link_to_relative_html_file_path(song[\"link\"]),\n",
    "                root_path=fu.relative_to_absolute_path(\n",
    "                    \"GetAllSongHTML/song_list/raw_html\",\n",
    "                    root_path=root_directory\n",
    "                )\n",
    "            )\n",
    "        )\n",
    "    ]\n",
    "\n",
    "    n_song_processed = 0\n",
    "    thread_timeout = 60\n",
    "\n",
    "    while(len(song_links) > 0):\n",
    "        current_song_per_batch = song_per_batch if song_per_batch <= len(song_links) else len(song_links)\n",
    "        song_batch = [song_links.pop(0) for _ in range(current_song_per_batch)]\n",
    "\n",
    "        threadBatch = []\n",
    "\n",
    "        for webdriver_pool_index in range(len(webdriver_pool)):\n",
    "            if webdriver_pool[webdriver_pool_index][\"browser_status\"] == BrowserStatus.AVAILABLE:\n",
    "                threadBatch.insert(\n",
    "                    0,\n",
    "                    threading.Thread(\n",
    "                        target=get_raw_html_with_webdriver_from_pool,\n",
    "                        args=(song_batch.pop(0)[\"link\"], webdriver_pool_index, )\n",
    "                    )\n",
    "                )\n",
    "\n",
    "                threadBatch[0].start()\n",
    "            elif webdriver_pool_index == len(webdriver_pool) - 1:\n",
    "                print(f\"All webdrivers are busy\")\n",
    "                break\n",
    "        \n",
    "        for i in range(len(threadBatch)):\n",
    "            threadBatch[i].join(thread_timeout)\n",
    "\n",
    "\n",
    "        anti_rate_limiting_waited_time = 0\n",
    "        total_time_to_wait = 30\n",
    "\n",
    "        for i in range(total_time_to_wait):\n",
    "            time.sleep(1)\n",
    "            anti_rate_limiting_waited_time += 1\n",
    "            print(f\"Waited {anti_rate_limiting_waited_time} of {total_time_to_wait} seconds\")\n",
    "        \n",
    "        threadBatch.clear()\n",
    "\n",
    "        n_song_processed += current_song_per_batch\n",
    "        print(f\"{n_song_processed} songs processed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n",
      "Webdriver on port 40000 started successfully\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n",
      "Webdriver on port 40001 started successfully\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Webdriver on port 40002 started successfully\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Opening Edge in remote debugging mode on port 40003...\n",
      "Webdriver on port 40003 started successfully\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n",
      "Webdriver on port 40004 started successfully\n"
     ]
    }
   ],
   "source": [
    "song_per_batch = 5\n",
    "init_webdriver_pool(n_webdriver=song_per_batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.hooktheory.com/theorytab/view/akira-yamaoka/promise: processing...\n",
      "https://www.hooktheory.com/theorytab/view/alstroemeria-records/bad-apple: processing...\n",
      "https://www.hooktheory.com/theorytab/view/alstroemeria-records/lovelight---bad-apple: processing...\n",
      "https://www.hooktheory.com/theorytab/view/amber-island/big-blue-bubble: processing...\n",
      "https://www.hooktheory.com/theorytab/view/amber-run/i-found: processing...\n",
      "Scrolled: 1 time(s)\n",
      "Scrolled: 1 time(s)\n",
      "Scrolled: 1 time(s)\n",
      "Scrolled: 2 time(s)\n",
      "Scrolled: 1 time(s)\n",
      "Scrolled: 1 time(s)\n",
      "Scrolled: 3 time(s)\n",
      "Scrolled: 2 time(s)\n",
      "Scrolled: 3 time(s)\n",
      "Scrolled: 2 time(s)\n",
      "Scrolled: 4 time(s)\n",
      "Scrolled: 4 time(s)\n",
      "Scrolled: 2 time(s)\n",
      "Scrolled: 5 time(s)\n",
      "https://www.hooktheory.com/theorytab/view/amber-island/big-blue-bubble: done\n",
      "https://www.hooktheory.com/theorytab/view/amber-run/i-found: done\n",
      "Scrolled: 3 time(s)\n",
      "Scrolled: 3 time(s)\n",
      "Scrolled: 4 time(s)\n",
      "Scrolled: 4 time(s)\n",
      "Scrolled: 5 time(s)\n",
      "Scrolled: 5 time(s)\n",
      "Scrolled: 6 time(s)\n",
      "Scrolled: 6 time(s)\n",
      "Scrolled: 7 time(s)\n",
      "Scrolled: 2 time(s)\n",
      "Scrolled: 7 time(s)\n",
      "Scrolled: 8 time(s)\n",
      "Scrolled: 8 time(s)\n",
      "Scrolled: 9 time(s)\n",
      "Scrolled: 10 time(s)\n",
      "Scrolled: 3 time(s)\n",
      "Scrolled: 9 time(s)\n",
      "Scrolled: 11 time(s)\n",
      "Scrolled: 4 time(s)\n",
      "Scrolled: 10 time(s)\n",
      "Scrolled: 12 time(s)\n",
      "Scrolled: 5 time(s)\n",
      "Scrolled: 11 time(s)\n",
      "Scrolled: 13 time(s)\n",
      "Scrolled: 6 time(s)\n",
      "Scrolled: 12 time(s)\n",
      "Scrolled: 14 time(s)\n",
      "Scrolled: 7 time(s)\n",
      "Scrolled: 13 time(s)\n",
      "Scrolled: 15 time(s)\n",
      "Scrolled: 8 time(s)\n",
      "Scrolled: 14 time(s)\n",
      "Scrolled: 16 time(s)\n",
      "Scrolled: 9 time(s)\n",
      "Scrolled: 15 time(s)\n",
      "Scrolled: 17 time(s)\n",
      "Scrolled: 10 time(s)\n",
      "Scrolled: 16 time(s)\n",
      "Scrolled: 18 time(s)\n",
      "Scrolled: 11 time(s)\n",
      "Scrolled: 17 time(s)\n",
      "Scrolled: 19 time(s)\n",
      "Scrolled: 12 time(s)\n",
      "Scrolled: 18 time(s)\n",
      "Scrolled: 20 time(s)\n",
      "Scrolled: 13 time(s)\n",
      "Scrolled: 19 time(s)\n",
      "Scrolled: 21 time(s)\n",
      "Scrolled: 14 time(s)\n",
      "Scrolled: 20 time(s)Scrolled: 22 time(s)\n",
      "\n",
      "Scrolled: 15 time(s)\n",
      "Scrolled: 21 time(s)Scrolled: 23 time(s)\n",
      "\n",
      "Scrolled: 16 time(s)\n",
      "Scrolled: 22 time(s)\n",
      "Scrolled: 24 time(s)\n",
      "Scrolled: 17 time(s)\n",
      "Scrolled: 23 time(s)\n",
      "Scrolled: 18 time(s)\n",
      "Scrolled: 25 time(s)\n",
      "Scrolled: 19 time(s)\n",
      "Max scroll times reached\n",
      "Killing process running on port 40000...\n",
      "Scrolled: 24 time(s)\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40000...\n",
      "Scrolled: 20 time(s)\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40000...Killing process running on port 40000...\n",
      "\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n",
      "Webdriver on port 40000 started successfully\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n",
      "Failed to start webdriver on port 40000. Retrying...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n",
      "Killing process running on port 40000...\n",
      "Webdriver on port 40000 is already started\n",
      "Failed to start webdriver on port 40000. Retrying...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-772 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 50, in __init__\n",
      "    self.service.start()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 102, in start\n",
      "    self.assert_process_still_running()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 115, in assert_process_still_running\n",
      "    raise WebDriverException(f\"Service {self._path} unexpectedly exited. Status code was: {return_code}\")\n",
      "selenium.common.exceptions.WebDriverException: Message: Service C:\\Users\\Vector A\\.cache\\selenium\\msedgedriver\\win64\\121.0.2277.98\\msedgedriver.exe unexpectedly exited. Status code was: 1\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-763 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "    response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "    return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Webdriver on port 40002 is already started\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Opening Edge in remote debugging mode on port 40003...\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n",
      "Webdriver on port 40003 is already started\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Webdriver on port 40003 is already started\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Webdriver on port 40004 is already started\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Waited 1 of 30 seconds\n",
      "Waited 2 of 30 seconds\n",
      "Waited 3 of 30 seconds\n",
      "Waited 4 of 30 seconds\n",
      "Waited 5 of 30 seconds\n",
      "Waited 6 of 30 seconds\n",
      "Waited 7 of 30 seconds\n",
      "Waited 8 of 30 seconds\n",
      "Waited 9 of 30 seconds\n",
      "Waited 10 of 30 seconds\n",
      "Waited 11 of 30 seconds\n",
      "Waited 12 of 30 seconds\n",
      "Waited 13 of 30 seconds\n",
      "Waited 14 of 30 seconds\n",
      "Waited 15 of 30 seconds\n",
      "Waited 16 of 30 seconds\n",
      "Waited 17 of 30 seconds\n",
      "Waited 18 of 30 seconds\n",
      "Waited 19 of 30 seconds\n",
      "Waited 20 of 30 seconds\n",
      "Waited 21 of 30 seconds\n",
      "Waited 22 of 30 seconds\n",
      "Waited 23 of 30 seconds\n",
      "Waited 24 of 30 seconds\n",
      "Waited 25 of 30 seconds\n",
      "Waited 26 of 30 seconds\n",
      "Waited 27 of 30 seconds\n",
      "Waited 28 of 30 seconds\n",
      "Waited 29 of 30 seconds\n",
      "Waited 30 of 30 seconds\n",
      "5 songs processed\n",
      "https://www.hooktheory.com/theorytab/view/amber/ocean: processing...\n",
      "https://www.hooktheory.com/theorytab/view/ambika/baba-nam-kevalam: processing...\n",
      "https://www.hooktheory.com/theorytab/view/amelie-p/basket-case: processing...\n",
      "https://www.hooktheory.com/theorytab/view/america/a-horse-with-no-name: processing...\n",
      "Failed to get https://www.hooktheory.com/theorytab/view/amelie-p/basket-case\n",
      "Killing process running on port 40000...\n",
      "https://www.hooktheory.com/theorytab/view/america/sister-golden-hair: processing...\n",
      "All webdrivers are busy\n",
      "Failed to get https://www.hooktheory.com/theorytab/view/america/a-horse-with-no-name\n",
      "Killing process running on port 40000...\n",
      "Failed to get https://www.hooktheory.com/theorytab/view/america/sister-golden-hair\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40002...Killing process running on port 40002...\n",
      "\n",
      "Killing process running on port 40002...\n",
      "Failed to get https://www.hooktheory.com/theorytab/view/ambika/baba-nam-kevalam\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40000...Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n",
      "Webdriver on port 40000 started successfully\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n",
      "Failed to start webdriver on port 40000. Retrying...\n",
      "Killing process running on port 40001...\n",
      "Failed to start webdriver on port 40000. Retrying...\n",
      "Killing process running on port 40001...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-798 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "    response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "    return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n",
      "Killing process running on port 40000...\n",
      "Webdriver on port 40000 is already started\n",
      "Failed to start webdriver on port 40000. Retrying...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-928 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 50, in __init__\n",
      "    self.service.start()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 102, in start\n",
      "    self.assert_process_still_running()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 115, in assert_process_still_running\n",
      "    raise WebDriverException(f\"Service {self._path} unexpectedly exited. Status code was: {return_code}\")\n",
      "selenium.common.exceptions.WebDriverException: Message: Service C:\\Users\\Vector A\\.cache\\selenium\\msedgedriver\\win64\\121.0.2277.98\\msedgedriver.exe unexpectedly exited. Status code was: 1\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Failed to get https://www.hooktheory.com/theorytab/view/amber/ocean\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Opening Edge in remote debugging mode on port 40003...\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Webdriver on port 40003 is already started\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Webdriver on port 40003 is already started\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Webdriver on port 40003 is already started\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Killing process running on port 40005...\n",
      "Killing process running on port 40005...\n",
      "Opening Edge in remote debugging mode on port 40005...\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Killing process running on port 40005...\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Killing process running on port 40005...\n",
      "Killing process running on port 40005...\n",
      "Opening Edge in remote debugging mode on port 40005...\n",
      "Killing process running on port 40005...\n",
      "Opening Edge in remote debugging mode on port 40005...\n",
      "Webdriver on port 40004 started successfully\n",
      "Killing process running on port 40005...\n",
      "Killing process running on port 40005...\n",
      "Opening Edge in remote debugging mode on port 40005...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Exception in thread Thread-976 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "Exception in thread Thread-979 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "Exception in thread Thread-963 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "Thread-1076 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "Exception in thread Thread-1063 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "Exception in thread Thread-1079 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "Exception in thread Thread-1090 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "    response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "    response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "    response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "    "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to start webdriver on port 40000. Retrying...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "    response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "        response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "        httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to start webdriver on port 40005. Retrying...\n",
      "Killing process running on port 40006...\n",
      "Failed to start webdriver on port 40005. Retrying...\n",
      "Killing process running on port 40006...\n",
      "Failed to start webdriver on port 40005. Retrying...\n",
      "Killing process running on port 40006...\n",
      "Failed to start webdriver on port 40005. Retrying...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Opening Edge in remote debugging mode on port 40006...\n",
      "Opening Edge in remote debugging mode on port 40006...Opening Edge in remote debugging mode on port 40006...\n",
      "\n",
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Webdriver on port 40002 is already started\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Webdriver on port 40003 is already started\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Webdriver on port 40004 is already started\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Killing process running on port 40005...\n",
      "Killing process running on port 40005...\n",
      "Opening Edge in remote debugging mode on port 40005...\n",
      "Failed to start webdriver on port 40006. Retrying...\n",
      "Failed to start webdriver on port 40006. Retrying...\n",
      "Failed to start webdriver on port 40006. Retrying...\n",
      "Failed to start webdriver on port 40005. Retrying...\n",
      "Failed to start webdriver on port 40006. Retrying...\n",
      "Waited 1 of 30 seconds\n",
      "Waited 2 of 30 seconds\n",
      "Waited 3 of 30 seconds\n",
      "Waited 4 of 30 seconds\n",
      "Waited 5 of 30 seconds\n",
      "Opening Edge in remote debugging mode on port 40006...\n",
      "Waited 6 of 30 seconds\n",
      "Waited 7 of 30 seconds\n",
      "Waited 8 of 30 seconds\n",
      "Waited 9 of 30 seconds\n",
      "Waited 10 of 30 seconds\n",
      "Waited 11 of 30 seconds\n",
      "Waited 12 of 30 seconds\n",
      "Waited 13 of 30 seconds\n",
      "Waited 14 of 30 seconds\n",
      "Waited 15 of 30 seconds\n",
      "Waited 16 of 30 seconds\n",
      "Waited 17 of 30 seconds\n",
      "Waited 18 of 30 seconds\n",
      "Waited 19 of 30 seconds\n",
      "Waited 20 of 30 seconds\n",
      "Waited 21 of 30 seconds\n",
      "Waited 22 of 30 seconds\n",
      "Waited 23 of 30 seconds\n",
      "Waited 24 of 30 seconds\n",
      "Waited 25 of 30 seconds\n",
      "Waited 26 of 30 seconds\n",
      "Waited 27 of 30 seconds\n",
      "Waited 28 of 30 seconds\n",
      "Waited 29 of 30 seconds\n",
      "Waited 30 of 30 seconds\n",
      "10 songs processed\n",
      "https://www.hooktheory.com/theorytab/view/america/tin-man: processing...\n",
      "https://www.hooktheory.com/theorytab/view/america/you-can-do-magic: processing...\n",
      "https://www.hooktheory.com/theorytab/view/american-authors/best-day-of-my-life: processing...\n",
      "https://www.hooktheory.com/theorytab/view/american-football/never-meant: processing...\n",
      "https://www.hooktheory.com/theorytab/view/american-football/stay-home: processing...\n",
      "Failed to get https://www.hooktheory.com/theorytab/view/american-authors/best-day-of-my-life\n",
      "Killing process running on port 40003...\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "pop from empty list",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[48], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mget_song_raw_html_by_alphabet\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43ma\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msong_per_batch\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[46], line 33\u001b[0m, in \u001b[0;36mget_song_raw_html_by_alphabet\u001b[1;34m(letter, song_per_batch)\u001b[0m\n\u001b[0;32m     27\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m webdriver_pool_index \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(webdriver_pool)):\n\u001b[0;32m     28\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m webdriver_pool[webdriver_pool_index][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbrowser_status\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m==\u001b[39m BrowserStatus\u001b[38;5;241m.\u001b[39mAVAILABLE:\n\u001b[0;32m     29\u001b[0m         threadBatch\u001b[38;5;241m.\u001b[39minsert(\n\u001b[0;32m     30\u001b[0m             \u001b[38;5;241m0\u001b[39m,\n\u001b[0;32m     31\u001b[0m             threading\u001b[38;5;241m.\u001b[39mThread(\n\u001b[0;32m     32\u001b[0m                 target\u001b[38;5;241m=\u001b[39mget_raw_html_with_webdriver_from_pool,\n\u001b[1;32m---> 33\u001b[0m                 args\u001b[38;5;241m=\u001b[39m(\u001b[43msong_batch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpop\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlink\u001b[39m\u001b[38;5;124m\"\u001b[39m], webdriver_pool_index, )\n\u001b[0;32m     34\u001b[0m             )\n\u001b[0;32m     35\u001b[0m         )\n\u001b[0;32m     37\u001b[0m         threadBatch[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mstart()\n\u001b[0;32m     38\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m webdriver_pool_index \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mlen\u001b[39m(webdriver_pool) \u001b[38;5;241m-\u001b[39m \u001b[38;5;241m1\u001b[39m:\n",
      "\u001b[1;31mIndexError\u001b[0m: pop from empty list"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40001...\n",
      "Failed to get https://www.hooktheory.com/theorytab/view/american-football/never-meant\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40004...\n",
      "Failed to get https://www.hooktheory.com/theorytab/view/american-football/stay-home\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40005...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40005...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n",
      "Webdriver on port 40000 started successfully\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n",
      "Failed to start webdriver on port 40000. Retrying...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n",
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Opening Edge in remote debugging mode on port 40003...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Opening Edge in remote debugging mode on port 40003...\n",
      "Failed to get https://www.hooktheory.com/theorytab/view/america/you-can-do-magicFailed to get https://www.hooktheory.com/theorytab/view/america/tin-man\n",
      "Killing process running on port 40000...\n",
      "\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40000...\n",
      "Killing process running on port 40000...\n",
      "Opening Edge in remote debugging mode on port 40000...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-1334 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 50, in __init__\n",
      "    self.service.start()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 102, in start\n",
      "    self.assert_process_still_running()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 115, in assert_process_still_running\n",
      "    raise WebDriverException(f\"Service {self._path} unexpectedly exited. Status code was: {return_code}\")\n",
      "selenium.common.exceptions.WebDriverException: Message: Service C:\\Users\\Vector A\\.cache\\selenium\\msedgedriver\\win64\\121.0.2277.98\\msedgedriver.exe unexpectedly exited. Status code was: 1\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Webdriver on port 40000 started successfully\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-1307 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 61, in __init__\n",
      "    super().__init__(command_executor=executor, options=options)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 208, in __init__\n",
      "    self.start_session(capabilities)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 292, in start_session\n",
      "    response = self.execute(Command.NEW_SESSION, caps)[\"value\"]\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\", line 345, in execute\n",
      "    response = self.command_executor.execute(driver_command, params)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 302, in execute\n",
      "    return self._request(command_info[0], url, body=data)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\remote\\remote_connection.py\", line 322, in _request\n",
      "    response = self._conn.request(method, url, body=body, headers=headers)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 78, in request\n",
      "    return self.request_encode_body(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\request.py\", line 170, in request_encode_body\n",
      "    return self.urlopen(method, url, **extra_kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\poolmanager.py\", line 376, in urlopen\n",
      "    response = conn.urlopen(method, u.request_uri, **kw)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 787, in urlopen\n",
      "    retries = retries.increment(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\util\\retry.py\", line 550, in increment\n",
      "    raise six.reraise(type(error), error, _stacktrace)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\packages\\six.py\", line 769, in reraise\n",
      "    raise value.with_traceback(tb)\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 703, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 449, in _make_request\n",
      "    six.raise_from(e, None)\n",
      "  File \"<string>\", line 3, in raise_from\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\urllib3\\connectionpool.py\", line 444, in _make_request\n",
      "    httplib_response = conn.getresponse()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 1374, in getresponse\n",
      "    response.begin()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 318, in begin\n",
      "    version, status, reason = self._read_status()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\http\\client.py\", line 279, in _read_status\n",
      "    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\socket.py\", line 705, in readinto\n",
      "    return self._sock.recv_into(b)\n",
      "urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-1376 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 50, in __init__\n",
      "    self.service.start()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 102, in start\n",
      "    self.assert_process_still_running()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 115, in assert_process_still_running\n",
      "    raise WebDriverException(f\"Service {self._path} unexpectedly exited. Status code was: {return_code}\")\n",
      "selenium.common.exceptions.WebDriverException: Message: Service C:\\Users\\Vector A\\.cache\\selenium\\msedgedriver\\win64\\121.0.2277.98\\msedgedriver.exe unexpectedly exited. Status code was: 1\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Killing process running on port 40000...\n",
      "Webdriver on port 40000 is already started\n",
      "Webdriver on port 40000 started successfully\n",
      "Killing process running on port 40001...\n",
      "Killing process running on port 40001...\n",
      "Opening Edge in remote debugging mode on port 40001...\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Killing process running on port 40005...\n",
      "Killing process running on port 40005...\n",
      "Opening Edge in remote debugging mode on port 40005...\n",
      "Webdriver on port 40001 started successfully\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Webdriver on port 40001 started successfully\n",
      "Killing process running on port 40002...\n",
      "Killing process running on port 40002...\n",
      "Opening Edge in remote debugging mode on port 40002...\n",
      "Failed to start webdriver on port 40005. Retrying...\n",
      "Killing process running on port 40006...\n",
      "Killing process running on port 40006...\n",
      "Opening Edge in remote debugging mode on port 40006...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Opening Edge in remote debugging mode on port 40003...\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Killing process running on port 40003...\n",
      "Opening Edge in remote debugging mode on port 40003...\n",
      "Failed to start webdriver on port 40006. Retrying...\n",
      "Killing process running on port 40007...\n",
      "Killing process running on port 40007...\n",
      "Opening Edge in remote debugging mode on port 40007...\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n",
      "Killing process running on port 40004...\n",
      "Opening Edge in remote debugging mode on port 40004...\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Killing process running on port 40005...\n",
      "Failed to start webdriver on port 40007. Retrying...\n",
      "Killing process running on port 40008...\n",
      "Killing process running on port 40008...\n",
      "Opening Edge in remote debugging mode on port 40008...\n",
      "Webdriver on port 40005 is already started\n",
      "Failed to start webdriver on port 40005. Retrying...\n",
      "Killing process running on port 40006...\n",
      "Webdriver on port 40006 is already started\n",
      "Failed to start webdriver on port 40006. Retrying...\n",
      "Killing process running on port 40007...\n",
      "Webdriver on port 40004 is already started\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Killing process running on port 40005...\n",
      "Webdriver on port 40005 is already started\n",
      "Failed to start webdriver on port 40005. Retrying...\n",
      "Killing process running on port 40006...\n",
      "Webdriver on port 40006 is already started\n",
      "Failed to start webdriver on port 40006. Retrying...\n",
      "Killing process running on port 40007...\n",
      "Webdriver on port 40007 is already started\n",
      "Failed to start webdriver on port 40007. Retrying...\n",
      "Killing process running on port 40008...\n",
      "Webdriver on port 40007 is already started\n",
      "Failed to start webdriver on port 40007. Retrying...\n",
      "Killing process running on port 40008...\n",
      "Killing process running on port 40008...\n",
      "Killing process running on port 40008...\n",
      "Opening Edge in remote debugging mode on port 40008...\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Killing process running on port 40005...\n",
      "Webdriver on port 40005 is already started\n",
      "Failed to start webdriver on port 40005. Retrying...\n",
      "Opening Edge in remote debugging mode on port 40008...\n",
      "Killing process running on port 40000...\n",
      "Webdriver on port 40000 is already started\n",
      "Failed to start webdriver on port 40000. Retrying...\n",
      "Killing process running on port 40001...\n",
      "Webdriver on port 40001 is already started\n",
      "Failed to start webdriver on port 40001. Retrying...\n",
      "Killing process running on port 40002...\n",
      "Webdriver on port 40002 is already started\n",
      "Failed to start webdriver on port 40002. Retrying...\n",
      "Killing process running on port 40003...\n",
      "Webdriver on port 40003 is already started\n",
      "Failed to start webdriver on port 40003. Retrying...\n",
      "Killing process running on port 40004...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-1468 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 50, in __init__\n",
      "    self.service.start()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 102, in start\n",
      "    self.assert_process_still_running()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 115, in assert_process_still_running\n",
      "    raise WebDriverException(f\"Service {self._path} unexpectedly exited. Status code was: {return_code}\")\n",
      "selenium.common.exceptions.WebDriverException: Message: Service C:\\Users\\Vector A\\.cache\\selenium\\msedgedriver\\win64\\121.0.2277.98\\msedgedriver.exe unexpectedly exited. Status code was: 1\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Webdriver on port 40004 is already started\n",
      "Failed to start webdriver on port 40004. Retrying...\n",
      "Killing process running on port 40005...\n",
      "Webdriver on port 40005 is already started\n",
      "Failed to start webdriver on port 40005. Retrying...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-1504 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 50, in __init__\n",
      "    self.service.start()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 102, in start\n",
      "    self.assert_process_still_running()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 115, in assert_process_still_running\n",
      "    raise WebDriverException(f\"Service {self._path} unexpectedly exited. Status code was: {return_code}\")\n",
      "selenium.common.exceptions.WebDriverException: Message: Service C:\\Users\\Vector A\\.cache\\selenium\\msedgedriver\\win64\\121.0.2277.98\\msedgedriver.exe unexpectedly exited. Status code was: 1\n",
      "\n",
      "Exception in thread Thread-1509 (start_webdriver):\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 1016, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\threading.py\", line 953, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2602687824.py\", line 15, in start_webdriver\n",
      "  File \"C:\\Users\\Vector A\\AppData\\Local\\Temp\\ipykernel_23920\\2828608030.py\", line 10, in init_selenium_in_remote_debugging_mode\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\edge\\webdriver.py\", line 45, in __init__\n",
      "    super().__init__(\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\chromium\\webdriver.py\", line 50, in __init__\n",
      "    self.service.start()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 102, in start\n",
      "    self.assert_process_still_running()\n",
      "  File \"c:\\Users\\Vector A\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\selenium\\webdriver\\common\\service.py\", line 115, in assert_process_still_running\n",
      "    raise WebDriverException(f\"Service {self._path} unexpectedly exited. Status code was: {return_code}\")\n",
      "selenium.common.exceptions.WebDriverException: Message: Service C:\\Users\\Vector A\\.cache\\selenium\\msedgedriver\\win64\\121.0.2277.98\\msedgedriver.exe unexpectedly exited. Status code was: 1\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to start webdriver on port 40008. Retrying...\n",
      "Killing process running on port 40009...\n",
      "Killing process running on port 40009...\n",
      "Opening Edge in remote debugging mode on port 40009...\n",
      "Failed to start webdriver on port 40008. Retrying...\n",
      "Killing process running on port 40009...\n",
      "Failed to start webdriver on port 40008. Retrying...\n",
      "Killing process running on port 40009...\n",
      "Killing process running on port 40009...\n",
      "Opening Edge in remote debugging mode on port 40009...\n",
      "Killing process running on port 40009...\n",
      "Opening Edge in remote debugging mode on port 40009...\n",
      "Failed to start webdriver on port 40009. Retrying...\n",
      "Killing process running on port 40010...\n",
      "Killing process running on port 40010...\n",
      "Opening Edge in remote debugging mode on port 40010...\n",
      "Failed to start webdriver on port 40009. Retrying...\n",
      "Killing process running on port 40010...\n",
      "Failed to start webdriver on port 40009. Retrying...\n",
      "Killing process running on port 40010...\n",
      "Killing process running on port 40010...\n",
      "Killing process running on port 40010...\n",
      "Opening Edge in remote debugging mode on port 40010...\n",
      "Opening Edge in remote debugging mode on port 40010...\n",
      "Failed to start webdriver on port 40010. Retrying...\n",
      "Killing process running on port 40011...\n",
      "Killing process running on port 40011...\n",
      "Opening Edge in remote debugging mode on port 40011...\n",
      "Failed to start webdriver on port 40010. Retrying...\n",
      "Killing process running on port 40011...\n",
      "Failed to start webdriver on port 40010. Retrying...\n",
      "Killing process running on port 40011...\n",
      "Killing process running on port 40011...\n",
      "Opening Edge in remote debugging mode on port 40011...\n",
      "Killing process running on port 40011...\n",
      "Opening Edge in remote debugging mode on port 40011...\n",
      "Failed to start webdriver on port 40011. Retrying...\n",
      "Killing process running on port 40012...\n",
      "Killing process running on port 40012...\n",
      "Opening Edge in remote debugging mode on port 40012...\n",
      "Failed to start webdriver on port 40011. Retrying...\n",
      "Killing process running on port 40012...\n",
      "Failed to start webdriver on port 40011. Retrying...\n",
      "Killing process running on port 40012...\n",
      "Killing process running on port 40012...\n",
      "Killing process running on port 40012...\n",
      "Opening Edge in remote debugging mode on port 40012...\n",
      "Opening Edge in remote debugging mode on port 40012...\n",
      "Failed to start webdriver on port 40012. Retrying...\n",
      "Failed to start webdriver on port 40012. Retrying...\n",
      "Failed to start webdriver on port 40012. Retrying...\n"
     ]
    }
   ],
   "source": [
    "get_song_raw_html_by_alphabet('a', song_per_batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Kill all Edge instances\n",
    "for webdriver in webdriver_pool:\n",
    "    kill_process_running_on_port(webdriver[\"port\"], platform)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import platform\n",
    "\n",
    "# Lấy thông tin về hệ điều hành\n",
    "operating_system = platform.system()\n",
    "\n",
    "print(operating_system)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For debugging\n",
    "import os\n",
    "import platform as python_platform\n",
    "\n",
    "if python_platform.system() == \"Windows\":\n",
    "    os.system(\"taskkill /F /IM msedge.exe\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
